test/unittests/Connection_test.cpp

```diff
@@ -0,0 +1,134 @@
+    ConnectionCreationTest() = default;
+
+    std::unique_ptr<Dbus_interface_mock> mock_ { std::make_unique<Dbus_interface_mock>(Dbus_interface_mock()) };
+    sd_bus* STUB_ { reinterpret_cast<sd_bus*>(1) };
+};
+
+TEST_F(ConnectionCreationTest, DefaultConstructorNoThrowSystem)
+{
+    EXPECT_CALL(*mock_, sd_bus_open_system(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
+    EXPECT_CALL(*mock_, sd_bus_flush(_)).Times(1);
+    EXPECT_CALL(*mock_, sd_bus_flush_close_unref(_)).Times(1);
+    ASSERT_NO_THROW(sdbus::internal::Connection(BusType::eSystem, std::move(mock_)));
+}
+
+TEST_F(ConnectionCreationTest, DefaultConstructorNoThrowSession)
+{
+    EXPECT_CALL(*mock_, sd_bus_open_user(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
+    EXPECT_CALL(*mock_, sd_bus_flush(_)).Times(1);
+    EXPECT_CALL(*mock_, sd_bus_flush_close_unref(_)).Times(1);
+    ASSERT_NO_THROW(sdbus::internal::Connection(BusType::eSession, std::move(mock_)));
+}
+
+TEST_F(ConnectionCreationTest, DefaultConstructorBusOpenSystemThrow)
+{
+    EXPECT_CALL(*mock_, sd_bus_open_system(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(-1)));
+    ASSERT_THROW(sdbus::internal::Connection(BusType::eSystem, std::move(mock_)), sdbus::Error);
+}
+
+TEST_F(ConnectionCreationTest, DefaultConstructorBusOpenSessionThrow)
+{
+    EXPECT_CALL(*mock_, sd_bus_open_user(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(-1)));
+    ASSERT_THROW(sdbus::internal::Connection(BusType::eSession, std::move(mock_)), sdbus::Error);
+}
+
+TEST_F(ConnectionCreationTest, DefaultConstructorFlushSystemThrow)
+{
+    EXPECT_CALL(*mock_, sd_bus_open_system(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
+    EXPECT_CALL(*mock_, sd_bus_flush(_)).WillOnce(Return(-1));
+    EXPECT_CALL(*mock_, sd_bus_flush_close_unref(_)).Times(1);
+    ASSERT_THROW(sdbus::internal::Connection(BusType::eSystem, std::move(mock_)), sdbus::Error);
+}
+
+TEST_F(ConnectionCreationTest, DefaultConstructorFlushSessionThrow)
+{
+    EXPECT_CALL(*mock_, sd_bus_open_user(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
+    EXPECT_CALL(*mock_, sd_bus_flush(_)).WillOnce(Return(-1));
+    EXPECT_CALL(*mock_, sd_bus_flush_close_unref(_)).Times(1);
+    ASSERT_THROW(sdbus::internal::Connection(BusType::eSession, std::move(mock_)), sdbus::Error);
+}
+
+
+
+class ConnectionRequestTest : public ::testing::TestWithParam<BusType> {
+protected:
+    ConnectionRequestTest() = default;
+    void SetUp() override
+    {
+        switch (GetParam())
+        {
+            case BusType::eSystem:
+                EXPECT_CALL(*mock_, sd_bus_open_system(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
+                break;
+            case BusType::eSession:
+                EXPECT_CALL(*mock_, sd_bus_open_user(_)).WillOnce(DoAll(SetArgPointee<0>(STUB_), Return(1)));
+                break;
+            default:
+                break;
+        }
+        EXPECT_CALL(*mock_, sd_bus_flush(_)).Times(1);
```

Comment:
This two EXPECT_CALLs are actually ideal candidates to be replaced by ON_CALL. See Google Mock Cookbook, chapter [Knowing When to Expect](https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#knowing-when-to-expect)